### DISCUSSION TITLE
Escaping issues with MySQL datasource

### DISCUSSION THREAD

**Discussion by davidfb** [grafana/grafana#38343] [Data Sources]
**What happened**:

Variable substitution in MySQL datasource only escapes quote characters and pays no attention to all the other characters that should be escaped as stated in [https://dev.mysql.com/doc/refman/8.0/en/string-literals.html](https://dev.mysql.com/doc/refman/8.0/en/string-literals.html). This makes, for instance, the backslash character behave in unexpected manners.

**What you expected to happen**:

Full character substitution for variable values, giving proper treatment to backslashes and others, so that the generated SQL behaves as expected.

**How to reproduce it (as minimally and precisely as possible)**:

1. Construct a query by writing it's raw SQL and include on it a reference to a variable.
2. Give that variable a value that includes backslashes (e.g. 'test\with\backslashes').
3. The generated SQL will show that the variable value has been inserted as is, with no escaping of backslashes.
4. From MySQL point of view, those backslashes identify escaping sequences and will be ignored or interpreted depending on the character that comes after ('\w' will become simply 'w', whereas '\b' will be treated as a backspace character).

**Anything else we need to know?**:

As per documentation, MySQL behaves differently if the NO_BACKSLASH_ESCAPES variable is turned on (default is off). Current datasource escaping strategy (escaping only quotes by doubling them) would be ok when NO_BACKSLASH_ESCAPES is explicitly enabled, but normally this is not the case and a more complex strategy must be implemented. Provided there's these two different behaviours depending on the server configuration, a datasource configuration parameter could be added in order to choose between one escaping strategy or the other (let it be manually or automatically by querying the server configuration if possible).

Fixing the default variable format in the datasource so that it properly escapes all required characters won't prevent all possible problems, as there are many times when an alternative variable format, like 'regex', must be used in order to build the query, and those standard formats provided by Grafana core won't apply the necessary escaping. This is a problem already present with the current escaping strategy. In order to provide a full solution, on top of fixing datasource's default variable format in typescript, the creation of a Go $__escape macro that could be used to escape any string is suggested. An usage example could be:

```
SELECT
    time, metric, value
FROM
    items
WHERE
    metric REGEXP '$__escape(^MAIN\[${selection:regex}\]$)'
    AND $__unixEpochFilter(time);
```

If, for instance, selection variable had two values selected ('device(1)' and 'device(2)'), ${selection:regex} would be interpolated to `(device\(1\)|device\(2\))` and the result of calling the $__escape macro should be `^MAIN\\[(device\\(1\\)|device\\(2\\))\\]$` which correctly double escapes all backslashes so they remain meaningful when processing the regular expression in MySQL.

These problems are referenced in: https://community.grafana.com/t/variable-formatted-as-regex-mysql-double-escaping-of-backslashes/50311

**Environment**:
- Grafana version: 8.0.3
- Data source type & version: MySQL
- OS Grafana is installed on: Ubuntu
- User OS & Browser: Ubuntu / Google Chrome
- Grafana plugins: MySQL
- Others:

**Comment by ying-jeanne** [comment]
this should be a feature request instead. we need to see what should be done to make the use case work.

**Comment by pkolyvas** [comment]
We're in the process of moving Grafana feature requests to [GitHub Discussions](https://community.grafana.com/t/were-moving-grafana-feature-requests-to-github-discussions/50341/3). This issue is being moved to [Discussions](https://github.com/grafana/grafana/discussions) or has been moved from our issues list, depending on where you're seeing this message. 

 We hope to use the discussion thread to continue building use cases, and additional evidence based on engagement/voting, in order to help us prioritize this feature request later on down the road. 

As with all things Grafana, we value your input and participation in this project and appreciate your patience during this transitional period. Looking forward to the discussion!

**Comment by gabor** [comment]
we are moving these issues back to github issues, i created a github issue describing the problem at https://github.com/grafana/grafana/issues/78391 , if you think anything is missing or incorrect, please comment on the github issue. thanks!
